File: solutions/Virtual-Assistant/src/csharp/Microsoft.Bot.Solutions/Authentication/MultiProviderAuthDialog.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Solutions.Dialogs.BotResponseFormatters;
using Microsoft.Bot.Solutions.Extensions;
using Microsoft.Bot.Solutions.Middleware.Telemetry;
using Microsoft.Bot.Solutions.Resources;
using Microsoft.Bot.Solutions.Skills;

namespace Microsoft.Bot.Solutions.Authentication
{
    public class MultiProviderAuthDialog : ComponentDialog
    {
        private SkillConfigurationBase _skillConfiguration;
        private CommonResponseBuilder _responseBuilder = new CommonResponseBuilder();
        private string _selectedAuthType = string.Empty;

        public MultiProviderAuthDialog(SkillConfigurationBase skillConfiguration)
            : base(nameof(MultiProviderAuthDialog))
        {
            _skillConfiguration = skillConfiguration;

            if (_skillConfiguration.IsAuthenticatedSkill && !_skillConfiguration.AuthenticationConnections.Any())
            {
                throw new Exception("You must configure an authentication connection in your bot file before using this component.");
            }

            var auth = new WaterfallStep[]
            {
                PromptForProvider,
                PromptForAuth,
                HandleTokenResponse,
            };

            AddDialog(new WaterfallDialog(nameof(MultiProviderAuthDialog), auth));
            AddDialog(new ChoicePrompt(DialogIds.ProviderPrompt) { Style = ListStyle.SuggestedAction });

            foreach (var connection in skillConfiguration.AuthenticationConnections)
            {
                AddDialog(new OAuthPrompt(
                    connection.Key,
                    new OAuthPromptSettings
                    {
                        ConnectionName = connection.Key,
                        Title = CommonStrings.Login,
                        Text = string.Format(CommonStrings.LoginDescription, connection.Key),
                        Timeout = 30000,
                    },
                    AuthPromptValidator));
            }
        }

        private async Task<DialogTurnResult> PromptForProvider(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (_skillConfiguration.AuthenticationConnections.Count() == 1)
            {
                var result = _skillConfiguration.AuthenticationConnections.ElementAt(0).Key;
                return await stepContext.NextAsync(result);
            }
            else
            {
                var adapter = stepContext.Context.Adapter as BotFrameworkAdapter;
                var tokenStatusCollection = await adapter.GetTokenStatusAsync(stepContext.Context, stepContext.Context.Activity.From.Id);

                var matchingProviders = tokenStatusCollection.Where(p => p.HasToken && _skillConfiguration.AuthenticationConnections.Any(t => t.Key == p.ConnectionName)).ToList();

                if (matchingProviders.Count() == 1)
                {
                    var authType = matchingProviders[0].ConnectionName;
                    return await stepContext.NextAsync(authType);
                }
                else if (matchingProviders.Count() > 1)
                {
                    var choices = new List<Choice>();

                    foreach (var connection in matchingProviders)
                    {
                        choices.Add(new Choice()
                        {
                            Action = new CardAction(ActionTypes.ImBack, connection.ConnectionName, value: connection.ConnectionName),
                            Value = connection.ConnectionName,
                        });
                    }

                    return await stepContext.PromptAsync(DialogIds.ProviderPrompt, new PromptOptions
                    {
                        Prompt = stepContext.Context.Activity.CreateReply(CommonResponses.ConfiguredAuthProvidersPrompt),
                        Choices = choices,
                    });
                }
                else
                {
                    var choices = new List<Choice>();

                    foreach (var connection in _skillConfiguration.AuthenticationConnections)
                    {
                        choices.Add(new Choice()
                        {
                            Action = new CardAction(ActionTypes.ImBack, connection.Key, value: connection.Key),
                            Value = connection.Key,
                        });
                    }

                    return await stepContext.PromptAsync(DialogIds.ProviderPrompt, new PromptOptions
                    {
                        Prompt = stepContext.Context.Activity.CreateReply(CommonResponses.AuthProvidersPrompt),
                        Choices = choices,
                    });
                }
            }
        }

        private async Task<DialogTurnResult> PromptForAuth(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (stepContext.Result is string)
            {
                _selectedAuthType = stepContext.Result as string;
            }
            else if (stepContext.Result is FoundChoice)
            {
                var choice = stepContext.Result as FoundChoice;
                _selectedAuthType = choice.Value;
            }

            return await stepContext.PromptAsync(_selectedAuthType, new PromptOptions());
        }

        private async Task<DialogTurnResult> HandleTokenResponse(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tokenResponse = stepContext.Result as TokenResponse;
            if (tokenResponse != null && !string.IsNullOrWhiteSpace(tokenResponse.Token))
            {
                var result = await CreateProviderTokenResponse(stepContext.Context, tokenResponse);

                return await stepContext.EndDialogAsync(result);
            }
            else
            {
                TelemetryClient.TrackEventEx("TokenRetrievalFailure", stepContext.Context.Activity);

                stepContext.Context.Activity.CreateReply(CommonResponses.ErrorMessage_AuthFailure, null, new StringDictionary { { "authType", _selectedAuthType } });

                return new DialogTurnResult(DialogTurnStatus.Cancelled);
            }
        }

        private async Task<ProviderTokenResponse> CreateProviderTokenResponse(ITurnContext context, TokenResponse tokenResponse)
        {
            var adapter = context.Adapter as BotFrameworkAdapter;
            var tokens = await adapter.GetTokenStatusAsync(context, context.Activity.From.Id);
            var match = Array.Find(tokens, t => t.ConnectionName == tokenResponse.ConnectionName);

            return new ProviderTokenResponse
            {
                AuthenticationProvider = match.ServiceProviderDisplayName.GetAuthenticationProvider(),
                TokenResponse = tokenResponse,
            };
        }

        private Task<bool> AuthPromptValidator(PromptValidatorContext<TokenResponse> promptContext, CancellationToken cancellationToken)
        {
            var token = promptContext.Recognized.Value;
            if (token != null && !string.IsNullOrWhiteSpace(token.Token))
            {
                return Task.FromResult(true);
            }
            else
            {
                return Task.FromResult(false);
            }
        }

        private class DialogIds
        {
            public const string ProviderPrompt = "ProviderPrompt";
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/authentication/multiProviderAuthDialog.ts
import { BotFrameworkAdapter, TurnContext } from 'botbuilder';
import { Choice, ChoicePrompt, ComponentDialog, DialogTurnResult, DialogTurnStatus, FoundChoice,
    ListStyle, OAuthPrompt, PromptValidatorContext, WaterfallDialog, WaterfallStep, WaterfallStepContext } from 'botbuilder-dialogs';
// tslint:disable-next-line:no-submodule-imports
import { TokenStatus } from 'botframework-connector/lib/tokenApi/models';
import { ActionTypes, TokenResponse } from 'botframework-schema';
import { __ } from 'i18n';
import { TelemetryExtensions } from '../middleware';
import { CommonResponses } from '../resources';
import { ResponseManager } from '../responses/responseManager';
import { SkillConfigurationBase } from '../skills';
import { getAuthenticationProvider, IProviderTokenResponse } from './providerTokenResponse';

export class MultiProviderAuthDialog extends ComponentDialog {
    private readonly skillConfiguration: SkillConfigurationBase;
    private readonly responseManager: ResponseManager;
    private selectedAuthType!: string;

    constructor(skillConfiguration: SkillConfigurationBase) {
        super(MultiProviderAuthDialog.name);
        this.skillConfiguration = skillConfiguration;
        this.responseManager = new ResponseManager([CommonResponses], Array.from(this.skillConfiguration.localeConfigurations.keys()));
        if (this.skillConfiguration.isAuthenticatedSkill && !this.skillConfiguration.authenticationConnections) {
            throw new Error('You must configure an authentication connection in your bot file before using this component.');
        }

        const auth: WaterfallStep[] = [
            this.promptForProvider.bind(this),
            this.promptForAuth.bind(this),
            this.handleTokenResponse.bind(this)
        ];
        this.addDialog(new WaterfallDialog(MultiProviderAuthDialog.name, auth));

        const providerPrompt: ChoicePrompt = new ChoicePrompt(DialogIds.providerPrompt);
        providerPrompt.style = ListStyle.suggestedAction;
        this.addDialog(providerPrompt);

        Object.entries(this.skillConfiguration.authenticationConnections)
            .forEach((connection: [string, string]) => {
                const connectionKey: string = connection[0];
                const connectionPrompt: OAuthPrompt = new OAuthPrompt(
                    connectionKey,
                    {
                        connectionName: connectionKey,
                        title: __('commonStrings.login'),
                        text: __('commonStrings.loginDescription', connectionKey),
                        timeout: 30000
                    },
                    this.authPromptValidator);
                this.addDialog(connectionPrompt);
            });
    }

    private async promptForProvider(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        const authConnections: { key: string; value: string}[] = Object.entries(this.skillConfiguration.authenticationConnections)
            .map((entry: [string, string]) => {
                return { key: entry[0], value: entry[1] };
            });

        if (authConnections.length === 1) {
            const result: string = authConnections[0].key;

            return stepContext.next(result);

        } else {
            const adapter: BotFrameworkAdapter = <BotFrameworkAdapter> stepContext.context.adapter;
            // PENDING check adapter.getTokenStatusAsync
            const tokenStatusCollection: TokenStatus[] = [];

            const matchingProviders: TokenStatus[] = tokenStatusCollection
                .filter((val: TokenStatus) => {
                    return val.hasToken &&
                        Object.keys(this.skillConfiguration.authenticationConnections)
                        .some((k: string) => k === val.connectionName);
                });

            if (matchingProviders.length === 1) {
                const authType: string|undefined = matchingProviders[0].connectionName;

                if (!authType) {
                    throw new Error('Provider not found');
                }

                return stepContext.next(authType);

            } else if (matchingProviders.length > 1) {
                const choices: Choice[] = matchingProviders.map((connection: TokenStatus) => {

                    return {
                        action: {
                            type: ActionTypes.ImBack,
                            title: connection.connectionName || '',
                            value: connection.connectionName
                        },
                        value: connection.connectionName || ''
                    };
                });

                return stepContext.prompt(DialogIds.providerPrompt, {
                    prompt: this.responseManager.getResponse(CommonResponses.configuredAuthProvidersPrompt),
                    choices: choices
                });

            } else {
                const choices: Choice[] = Object.keys(this.skillConfiguration.authenticationConnections)
                    .map((key: string) => {

                        return {
                            action: {
                                type: ActionTypes.ImBack,
                                title: key,
                                value: key
                            },
                            value: key
                        };
                    });

                return stepContext.prompt(DialogIds.providerPrompt, {
                    prompt: this.responseManager.getResponse(CommonResponses.authProvidersPrompt),
                    choices: choices
                });
            }
        }
    }

    private promptForAuth(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        if (typeof stepContext.result === 'string') {
            this.selectedAuthType = stepContext.result;
        } else {
            const choice: FoundChoice = <FoundChoice> stepContext.result;
            if (choice && choice.value) {
                this.selectedAuthType = choice.value;
            }
        }

        return stepContext.prompt(this.selectedAuthType, {});
    }

    private async handleTokenResponse(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        const tokenResponse: TokenResponse = <TokenResponse> stepContext.result;

        if (tokenResponse && tokenResponse.token) {
            const result: IProviderTokenResponse = await this.createProviderTokenResponse(stepContext.context, tokenResponse);

            return stepContext.endDialog(result);

        } else {
            TelemetryExtensions.trackEventEx(this.telemetryClient, 'TokenRetrievalFailure', stepContext.context.activity);

            return { status: DialogTurnStatus.cancelled };
        }
    }

    private createProviderTokenResponse(context: TurnContext, tokenResponse: TokenResponse): Promise<IProviderTokenResponse> {
        const adapter: BotFrameworkAdapter = <BotFrameworkAdapter> context.adapter;
        // PENDING check adapter.getTokenStatusAsync
        const tokens: TokenStatus[] = [];
        const match: TokenStatus|undefined = tokens.find((t: TokenStatus) => t.connectionName === tokenResponse.connectionName);

        if (!match) {
            throw new Error('Token not found');
        }

        const response: IProviderTokenResponse = {
            authenticationProvider: getAuthenticationProvider(match.serviceProviderDisplayName || ''),
            tokenResponse: tokenResponse
        };

        return Promise.resolve(response);
    }

    private authPromptValidator(promptContext: PromptValidatorContext<TokenResponse>): Promise<boolean> {
        const token: TokenResponse|undefined = promptContext.recognized.value;
        const result: boolean = !!token && !!token.token;

        return Promise.resolve(result);
    }
}

namespace DialogIds {
    export const providerPrompt: string = 'ProviderPrompt';
}
